src/main/java/com/minecolonies/coremod/entity/ai/citizen/deliveryman/EntityAIWorkDeliveryman.java

```diff
@@ -13,12 +30,371 @@
+     * Check if the building still needs the item and if the required items are still in the warehouse.
+     * @return the next state to go to.
+     */
+    private AIState prepareDelivery()
+    {
+        final AbstractBuildingWorker ownBuilding = getOwnBuilding();
+        if(ownBuilding instanceof BuildingDeliveryman)
+        {
+            final AbstractBuilding buildingToDeliver = ((BuildingDeliveryman) ownBuilding).getBuildingToDeliver();
+            if(buildingToDeliver != null)
+            {
+                boolean ableToDeliver = wareHouse.getTileEntity().checkInWareHouse(buildingToDeliver);
+                if(!ableToDeliver)
+                {
+                    return START_WORKING;
+                }
+                itemsToDeliver = new ArrayList<>(buildingToDeliver.getNeededItems());
+                return GATHER_IN_WAREHOUSE;
+            }
+        }
+        return START_WORKING;
+    }
+
+    private AIState gatherItemsFromWareHouse()
+    {
+        final AbstractBuildingWorker ownBuilding = getOwnBuilding();
+        if (ownBuilding instanceof BuildingDeliveryman)
+        {
+            final AbstractBuilding buildingToDeliver = ((BuildingDeliveryman) ownBuilding).getBuildingToDeliver();
+            if (buildingToDeliver != null)
+            {
+                if(itemsToDeliver.isEmpty() && hasTools(buildingToDeliver))
+                {
+                    return DELIVERY;
+                }
+
+                gatherItems(buildingToDeliver);
+                return GATHER_IN_WAREHOUSE;
+            }
+        }
+        return START_WORKING;
+    }
+
+    /**
+     * Check if the deliveryman has all the tools to make the delivery.
+     * @param buildingToDeliver the building to deliver to.
+     * @return true if is ready to deliver.
+     */
+    private boolean hasTools(@NotNull final AbstractBuilding buildingToDeliver)
+    {
+        final String requiredTool = buildingToDeliver.getRequiredTool();
+        if(requiredTool.isEmpty())
+        {
+            return true;
+        }
+
+        if(requiredTool.equals(Utils.PICKAXE))
+        {
+            return InventoryUtils.isPickaxeInTileEntity(worker.getInventoryCitizen(), buildingToDeliver.getNeededPickaxeLevel());
+        }
+        return InventoryUtils.isToolInTileEntity(worker.getInventoryCitizen(), requiredTool);
+    }
+
+    /**
+     * Gather item from chest.
+     * Gathers only one stack of the item.
+     * @param buildingToDeliver
+     */
+    private void gatherItems(@NotNull final AbstractBuilding buildingToDeliver)
+    {
+        BlockPos position;
+        if(itemsToDeliver.isEmpty())
+        {
+            String tool = buildingToDeliver.getRequiredTool();
+            position = wareHouse.getTileEntity().getPositionOfChestWithTool(tool, Utils.PICKAXE.equals(tool) ? buildingToDeliver.getNeededPickaxeLevel() : -1);
+        }
+        else
+        {
+            ItemStack stack = itemsToDeliver.get(0);
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Local variable 'stack' could be declared final [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)